File: Ginger/src/net/pic4pic/ginger/entities/Locality.java
package net.pic4pic.ginger.entities;

import java.io.Serializable;

import com.google.gson.annotations.SerializedName;

public class Locality implements Serializable {

	private static final long serialVersionUID = 1L;

	@SerializedName("CountryCode")
	protected String countryCode;
	
	@SerializedName("Country")
	protected String country;
	
	@SerializedName("Region")
	protected String region;
	
	@SerializedName("SubRegion")
	protected String subRegion;
	
	@SerializedName("City")
	protected String city;
	
	@SerializedName("Neighborhood")
	protected String neighborhood;
	
	@SerializedName("ZipCode")
	protected String zipCode;
	
	@Override
	public String toString(){
		
		String address = "";
		
		if(this.city != null && this.city.length() > 0){
			
			if(address.length() > 0){
				address += ", ";
			}
			
			address += this.city;
		}
		
		if(this.region != null && this.region.length() > 0){
			
			if(address.length() > 0){
				address += ", ";
			}
			
			address += this.region;
		}
		
		if(this.country != null && this.country.length() > 0){
			
			if(address.length() > 0){
				address += ", ";
			}
			
			address += this.country;
		}
		
		if(this.zipCode != null && this.zipCode.length() > 0){
			
			if(address.length() > 0){
				address += ", ";
			}
			
			address += this.zipCode;
		}
		
		return address;
	}
	
	/**
	 * @return the city
	 */
	public String getCity() {
		return city;
	}

	/**
	 * @param city the city to set
	 */
	public void setCity(String city) {
		this.city = city;
	}

	/**
	 * @return the region
	 */
	public String getRegion() {
		return region;
	}

	/**
	 * @param region the region to set
	 */
	public void setRegion(String region) {
		this.region = region;
	}

	/**
	 * @return the country
	 */
	public String getCountry() {
		return country;
	}

	/**
	 * @param country the country to set
	 */
	public void setCountry(String country) {
		this.country = country;
	}

	/**
	 * @return the zipCode
	 */
	public String getZipCode() {
		return zipCode;
	}

	/**
	 * @param zipCode the zipCode to set
	 */
	public void setZipCode(String zipCode) {
		this.zipCode = zipCode;
	}

	/**
	 * @return the subRegion
	 */
	public String getSubRegion() {
		return subRegion;
	}

	/**
	 * @param subRegion the subRegion to set
	 */
	public void setSubRegion(String subRegion) {
		this.subRegion = subRegion;
	}

	/**
	 * @return the neighborhood
	 */
	public String getNeighborhood() {
		return neighborhood;
	}

	/**
	 * @param neighborhood the neighborhood to set
	 */
	public void setNeighborhood(String neighborhood) {
		this.neighborhood = neighborhood;
	}

	/**
	 * @return the countryCode
	 */
	public String getCountryCode() {
		return countryCode;
	}

	/**
	 * @param countryCode the countryCode to set
	 */
	public void setCountryCode(String countryCode) {
		this.countryCode = countryCode;
	}
}

File: Pic4Pic/Pic4Pic.ObjectModel/Entities/Locality.cs
using System;
using System.Globalization;
using System.Runtime.Serialization;

namespace Pic4Pic.ObjectModel
{
    [DataContract()]
    public class Locality : IVerifiable
    {
        [DataMember()]
        public String Country { get; set; }

        [DataMember()]
        public String CountryCode { get; set; }

        // a.k.a. State
        [DataMember()]
        public String Region { get; set; }

        // a.k.a. County
        [DataMember()]
        public String SubRegion { get; set; }

        [DataMember()]
        public String City { get; set; }

        [DataMember()]
        public String Neighborhood { get; set; }

        [DataMember()]
        public String ZipCode { get; set; }

        public override string ToString()
        {
            String address = "";

            if (!String.IsNullOrWhiteSpace(this.City))
            {
                if (!String.IsNullOrWhiteSpace(address))
                {
                    address += ", ";
                }

                address += this.City;
            }

            if (!String.IsNullOrWhiteSpace(this.Region))
            {
                if (!String.IsNullOrWhiteSpace(address))
                {
                    address += ", ";
                }

                address += this.Region;
            }

            if (!String.IsNullOrWhiteSpace(this.Country))
            {
                if (!String.IsNullOrWhiteSpace(address))
                {
                    address += ", ";
                }

                address += this.Country;
            }

            if (!String.IsNullOrWhiteSpace(this.ZipCode))
            {
                if (!String.IsNullOrWhiteSpace(address))
                {
                    address += ", ";
                }

                address += this.ZipCode;
            }

            return address;
        }

        public void Validate()
        {
            this.Validate(false);
        }

        public void Validate(bool checkZipCode)
        {
            if (String.IsNullOrWhiteSpace(this.Country))
            {
                throw new Pic4PicException("Country may not be empty");
            }

            if (String.IsNullOrWhiteSpace(this.Region))
            {
                throw new Pic4PicException("Region may not be empty");
            }

            if (String.IsNullOrWhiteSpace(this.City))
            {
                throw new Pic4PicException("City may not be empty");
            }

            if(checkZipCode)
            {
                if (String.IsNullOrWhiteSpace(this.ZipCode))
                {
                    throw new Pic4PicException("ZipCode may not be empty");
                }
            }
        }
    }
}
